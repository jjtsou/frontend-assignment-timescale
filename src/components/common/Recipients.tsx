import { Heading, Flex } from '@chakra-ui/react';
import { RecipientsByDomain } from '../../types/Recipient';
import { EmailList, EmailListItem } from '.';

type RecipientsProps = {
  header: string;
  recipients: RecipientsByDomain;
  handleSelectRecipient?: ({
    domain,
    email,
    isSelected,
  }: {
    domain: string;
    email: string;
    isSelected: boolean;
  }) => void;
  handleSelectDomain?: ({ domain }: { domain: string }) => void;
  isSelectable?: boolean;
  isGroup?: boolean;
};

const Recipients = ({
  header,
  recipients,
  handleSelectRecipient,
  handleSelectDomain,
  isSelectable,
  isGroup,
}: RecipientsProps): JSX.Element => {
  return (
    <Flex
      direction="column"
      justifyContent="start"
      alignItems="start"
      minH="45vh"
      p={20}
      pt={10}
      border="2px solid"
      borderColor="gray.400"
      rounded={5}
    >
      <Heading size="md" noOfLines={1} mb={5} color="gray.500">
        {header}
      </Heading>
      {Object.keys(recipients).map((domain) => (
        <EmailList
          key={`${domain}_list`}
          domain={domain}
          onSelectDomain={
            isSelectable ? () => handleSelectDomain?.({ domain }) : undefined
          }
          isGroup={isGroup || recipients[domain].length > 1}
          openByDefault={!!recipients[domain].length}
        >
          {recipients[domain].map(({ email, isSelected }, i) => {
            return (
              <EmailListItem
                key={`${email}_${i}_list_item`}
                isSelected={isSelected}
                value={email}
                onClick={
                  isSelectable
                    ? () =>
                        handleSelectRecipient?.({ domain, email, isSelected })
                    : undefined
                }
              />
            );
          })}
        </EmailList>
      ))}
    </Flex>
  );
};

export default Recipients;
