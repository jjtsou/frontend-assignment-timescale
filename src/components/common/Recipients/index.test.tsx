import { render } from '@testing-library/react';
import Recipients from '../Recipients';

const recipients = {
  'gmail.com': [
    {
      email: 'john@gmail.com',
      isSelected: false,
    },
    {
      email: 'alex@gmail.com',
      isSelected: true,
    },
  ],
  'timescale.com': [
    {
      email: 'mary@timescale.com',
      isSelected: false,
    },
  ],
};

test('renders correctly', () => {
  const { asFragment } = render(
    <Recipients header="Recipients" recipients={recipients} />
  );
  expect(asFragment()).toMatchSnapshot();
});
