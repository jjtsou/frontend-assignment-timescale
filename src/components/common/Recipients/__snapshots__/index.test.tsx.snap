// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-qbhc64"
  >
    <h2
      class="chakra-heading css-58v1or"
    >
      Recipients
    </h2>
    <div
      class="css-j7qwjs"
    >
      <div
        class="css-1m29ivc"
      >
        <button
          aria-label="Open collapsible panel"
          class="chakra-button css-9imh88"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-onkibi"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
              fill="currentColor"
            />
          </svg>
        </button>
        <p
          class="chakra-text chakra-text css-1efywoh"
        >
          gmail.com
        </p>
      </div>
      <div
        class="chakra-collapse"
        style="overflow: hidden; display: block; opacity: 1; height: auto;"
      >
        <div
          class="css-11qo4fr"
        >
          <p
            class="chakra-text chakra-text css-1k47jjv"
          >
            john@gmail.com
          </p>
          <p
            class="chakra-text chakra-text css-1pldbq1"
          >
            alex@gmail.com
          </p>
        </div>
      </div>
    </div>
    <p
      class="chakra-text chakra-text css-1k47jjv"
    >
      mary@timescale.com
    </p>
  </div>
</DocumentFragment>
`;
